From 2208ef4c56757429b1c6e17b6df932d28fb573d0 Mon Sep 17 00:00:00 2001
From: Rob Clark <rob@ti.com>
Date: Mon, 13 Sep 2010 19:04:47 -0500
Subject: [PATCH 22/26] video: more flexible video caps utility

Add gst_video_format_new_caps_simple() to allow for more flexible video
caps builder, which could be used for template caps and non-fixed caps.
---
 gst-libs/gst/video/video.c |  136 ++++++++++++++++++++++++++------------------
 gst-libs/gst/video/video.h |    2 +
 2 files changed, 83 insertions(+), 55 deletions(-)

diff --git a/gst-libs/gst/video/video.c b/gst-libs/gst/video/video.c
index 15474a2..d40efa7 100644
--- a/gst-libs/gst/video/video.c
+++ b/gst-libs/gst/video/video.c
@@ -585,15 +585,12 @@ gst_video_format_new_caps_interlaced (GstVideoFormat format,
 }
 
 /**
- * gst_video_format_new_caps_strided:
+ * gst_video_format_new_caps_simple:
  * @format: the #GstVideoFormat describing the raw video format
- * @width: width of video
- * @height: height of video
- * @rowstride: the rowstride (in bytes), or 0 if no rowstride
- * @framerate_n: numerator of frame rate
- * @framerate_d: denominator of frame rate
- * @par_n: numerator of pixel aspect ratio
- * @par_d: denominator of pixel aspect ratio
+ * @rowstride: 0 for unstrided, -1 for any stride (unfixed), or other
+ *     for fixed stride
+ * @fieldname: first field to set
+ * @...: additional arguments
  *
  * Creates a new #GstCaps object based on the parameters provided.
  *
@@ -602,25 +599,21 @@ gst_video_format_new_caps_interlaced (GstVideoFormat format,
  * Returns: a new #GstCaps object, or NULL if there was an error
  */
 GstCaps *
-gst_video_format_new_caps_strided (GstVideoFormat format,
-    int width, int height, int rowstride,
-    int framerate_n, int framerate_d, int par_n, int par_d)
+gst_video_format_new_caps_simple (GstVideoFormat format, int rowstride,
+    const char *fieldname, ...)
 {
-  GstCaps *caps = NULL;
+  va_list varargs;
+  GstStructure *s;
 
   g_return_val_if_fail (format != GST_VIDEO_FORMAT_UNKNOWN, NULL);
-  g_return_val_if_fail (width > 0 && height > 0, NULL);
+
 
   if (gst_video_format_is_yuv (format)) {
-    caps =
-        gst_caps_new_simple (rowstride ? "video/x-raw-yuv-strided" :
-        "video/x-raw-yuv", "format", GST_TYPE_FOURCC,
-        gst_video_format_to_fourcc (format), "width", G_TYPE_INT, width,
-        "height", G_TYPE_INT, height, "framerate", GST_TYPE_FRACTION,
-        framerate_n, framerate_d, "pixel-aspect-ratio", GST_TYPE_FRACTION,
-        par_n, par_d, NULL);
+    s = gst_structure_new (rowstride ?
+        "video/x-raw-yuv-strided" : "video/x-raw-yuv",
+        "format", GST_TYPE_FOURCC, gst_video_format_to_fourcc (format),
+        NULL);
   } else if (gst_video_format_is_rgb (format)) {
-    GstCaps *caps;
     int red_mask;
     int blue_mask;
     int green_mask;
@@ -674,15 +667,12 @@ gst_video_format_new_caps_strided (GstVideoFormat format,
       } else {
         mask = 0xff0000;
       }
-      red_mask =
-          mask >> (8 * gst_video_format_get_component_offset (format, 0,
-              width, height));
-      green_mask =
-          mask >> (8 * gst_video_format_get_component_offset (format, 1,
-              width, height));
-      blue_mask =
-          mask >> (8 * gst_video_format_get_component_offset (format, 2,
-              width, height));
+      red_mask = mask >>
+          (8 * gst_video_format_get_component_offset (format, 0, 1, 1));
+      green_mask = mask >>
+          (8 * gst_video_format_get_component_offset (format, 1, 1, 1));
+      blue_mask = mask >>
+          (8 * gst_video_format_get_component_offset (format, 2, 1, 1));
     } else if (bpp == 16) {
       switch (format) {
         case GST_VIDEO_FORMAT_RGB16:
@@ -713,19 +703,19 @@ gst_video_format_new_caps_strided (GstVideoFormat format,
       return NULL;
     }
 
-    caps =
-        gst_caps_new_simple (rowstride ? "video/x-raw-rgb-strided" :
-        "video/x-raw-rgb", "bpp", G_TYPE_INT, bpp, "depth", G_TYPE_INT, depth,
-        "endianness", G_TYPE_INT, G_BIG_ENDIAN, "red_mask", G_TYPE_INT,
-        red_mask, "green_mask", G_TYPE_INT, green_mask, "blue_mask", G_TYPE_INT,
-        blue_mask, "width", G_TYPE_INT, width, "height", G_TYPE_INT, height,
-        "framerate", GST_TYPE_FRACTION, framerate_n, framerate_d,
-        "pixel-aspect-ratio", GST_TYPE_FRACTION, par_n, par_d, NULL);
+    s = gst_structure_new (rowstride ?
+        "video/x-raw-rgb-strided" : "video/x-raw-rgb",
+        "bpp", G_TYPE_INT, bpp,
+        "depth", G_TYPE_INT, depth,
+        "endianness", G_TYPE_INT, G_BIG_ENDIAN,
+        "red_mask", G_TYPE_INT, red_mask,
+        "green_mask", G_TYPE_INT, green_mask,
+        "blue_mask", G_TYPE_INT, blue_mask,
+        NULL);
     if (have_alpha) {
-      alpha_mask =
-          mask >> (8 * gst_video_format_get_component_offset (format, 3,
-              width, height));
-      gst_caps_set_simple (caps, "alpha_mask", G_TYPE_INT, alpha_mask, NULL);
+      alpha_mask = mask >>
+          (8 * gst_video_format_get_component_offset (format, 3, 1, 1));
+      gst_structure_set (s, "alpha_mask", G_TYPE_INT, alpha_mask, NULL);
     }
   } else if (gst_video_format_is_gray (format)) {
     int bpp;
@@ -751,32 +741,68 @@ gst_video_format_new_caps_strided (GstVideoFormat format,
     }
 
     if (bpp > 8) {
-      caps = gst_caps_new_simple ("video/x-raw-gray",
+      s = gst_structure_new (rowstride ?
+          "video/x-raw-gray-strided" : "video/x-raw-gray",
           "bpp", G_TYPE_INT, bpp,
           "depth", G_TYPE_INT, depth,
-          "width", G_TYPE_INT, width,
-          "height", G_TYPE_INT, height,
-          "framerate", GST_TYPE_FRACTION, framerate_n, framerate_d,
-          "pixel-aspect-ratio", GST_TYPE_FRACTION, par_n, par_d, NULL);
+          NULL);
     } else {
-      caps = gst_caps_new_simple ("video/x-raw-gray",
+      s = gst_structure_new (rowstride ?
+          "video/x-raw-gray-strided" : "video/x-raw-gray",
           "bpp", G_TYPE_INT, bpp,
           "depth", G_TYPE_INT, depth,
           "endianness", G_TYPE_INT, G_BIG_ENDIAN,
-          "width", G_TYPE_INT, width,
-          "height", G_TYPE_INT, height,
-          "framerate", GST_TYPE_FRACTION, framerate_n, framerate_d,
-          "pixel-aspect-ratio", GST_TYPE_FRACTION, par_n, par_d, NULL);
+          NULL);
     }
   } else {
     return NULL;
   }
 
-  if (rowstride) {
-    gst_caps_set_simple (caps, "rowstride", G_TYPE_INT, rowstride, NULL);
+  if (rowstride > 0) {
+    gst_structure_set (s, "rowstride",
+        G_TYPE_INT, rowstride, NULL);
+  } else if (rowstride < 0) {
+    gst_structure_set (s, "rowstride",
+        GST_TYPE_INT_RANGE, 1, G_MAXINT, NULL);
   }
 
-  return caps;
+  va_start (varargs, fieldname);
+  gst_structure_set_valist (s, fieldname, varargs);
+  va_end (varargs);
+
+  return gst_caps_new_full (s, NULL);
+}
+
+/**
+ * gst_video_format_new_caps_strided:
+ * @format: the #GstVideoFormat describing the raw video format
+ * @width: width of video
+ * @height: height of video
+ * @rowstride: the rowstride (in bytes), or 0 if no rowstride
+ * @framerate_n: numerator of frame rate
+ * @framerate_d: denominator of frame rate
+ * @par_n: numerator of pixel aspect ratio
+ * @par_d: denominator of pixel aspect ratio
+ *
+ * Creates a new #GstCaps object based on the parameters provided.
+ *
+ * Since: ???
+ *
+ * Returns: a new #GstCaps object, or NULL if there was an error
+ */
+GstCaps *
+gst_video_format_new_caps_strided (GstVideoFormat format,
+    int width, int height, int rowstride,
+    int framerate_n, int framerate_d, int par_n, int par_d)
+{
+  g_return_val_if_fail (width > 0 && height > 0, NULL);
+
+  return gst_video_format_new_caps_simple (format, rowstride,
+      "width", G_TYPE_INT, width,
+      "height", G_TYPE_INT, height,
+      "framerate", GST_TYPE_FRACTION, framerate_n, framerate_d,
+      "pixel-aspect-ratio", GST_TYPE_FRACTION, par_n, par_d,
+      NULL);
 }
 
 /**
diff --git a/gst-libs/gst/video/video.h b/gst-libs/gst/video/video.h
index 367256e..8dc6b58 100644
--- a/gst-libs/gst/video/video.h
+++ b/gst-libs/gst/video/video.h
@@ -410,6 +410,8 @@ GstCaps * gst_video_format_new_caps_interlaced (GstVideoFormat format,
 GstCaps * gst_video_format_new_caps_strided (GstVideoFormat format,
     int width, int height, int rowstride,
     int framerate_n, int framerate_d, int par_n, int par_d);
+GstCaps * gst_video_format_new_caps_simple (GstVideoFormat format,
+    int rowstride, const char *fieldname, ...);
 GstVideoFormat gst_video_format_from_fourcc (guint32 fourcc);
 guint32 gst_video_format_to_fourcc (GstVideoFormat format);
 gboolean gst_video_format_is_rgb (GstVideoFormat format);
-- 
1.7.1

